### DISCUSSION TITLE
MNX identifier/file format

### DISCUSSION THREAD

**Discussion by mscuthbert** [w3c/mnx#305] [General]
Discussing file formats/extensions.  Related to #301 

Looking at:

* `.json` -- works fine for MNX but doesn't identify it as different from any other .json file  (like ".xml" for musicxml)
* `.mnx` -- the same JSON file but something you can click on to launch in a music editor.  (like ".musicxml" for MusicXML)
* `.mnxz` -- the JSON file compressed version (probably GZ but still undecided (like ".mxl" in MusicXML)

Also looking at putting the version (in some format) at a top-level tag that identifies the JSON as specifically MNX:

```
{
    "mnx": { "version": 27 }
}
```

Discussed at co-chairs meeting 21 June 2023.  Thoughts?

**Comment by samuelbradshaw** [comment]
Pros of .json (plain text)
– For developers: It's transparent / approachable to developers who haven't worked with music formats before
– For developers: Easy to store in a database as plain text
– For developers: Easy to open in any plain text editor

Pros of .mnx (plain text)
– For users: Can be set to open in music notation programs by default (unlike .json, which will open in a text editor by default)
– For developers: Easy to store in a database as plain text
– For developers: Easy to open in any plain text editor

Pros of .mnxz (compressed)
– For users: Can be set to open in music notation programs by default (unlike .json, which will open in a text editor by default)
– For users and developers: Takes up less space on a server, computer, or mobile device
– For users: Faster and smaller download

I lean towards:
- Compressed .mnxz as the primary user-facing format, because of the benefits for users.
- Don't have two different MNX extensions in the specification (this is something that was confusing in MusicXML).
- Make it easy for publishers to work internally with MNX in plain text by recommending that music notation programs support two import and export options: .mnxz for distribution to users, and .json for developer/publisher use. The simple .json file extension (rather than something like .mnx) will make it clear that it's not packaged for users.

**Reply by shoogle** [reply]
Good arguments. I agree the compressed format should be the default choice and needs to have a dedicated file extension.

Regarding the plain text format, I would add these points:

Pros of .json (plain text)
– Syntax highlighting in text editors enabled by default.

Cons of .json (plain text)
– Music notation programs would not be able to open the plain text format at all, or if they can, they would display all JSON files in their *File>Open* dialog rather than just the JSON files that are actually MNX.

For this latter reason, I would lean towards using a special extension even for the plain text format.

**Comment by shoogle** [comment]
I think we should encourage the compressed format as the primary choice for end users. To that end, why not use `.mnx` as the extension for the compressed format, and then pick a different extension for the plain text format, such as `.json` or `.mnxj`?

The compressed format has the advantage of being smaller and potentially able to include multiple files in the container (e.g. images, etc.), hence I believe it should be the default option for most users. The plain text format is mainly useful to developers.
